src/Microsoft.OData.Core/ODataBatchReader.cs

```diff
@@ -270,274 +263,25 @@ private Task<bool> ReadAsynchronously()
         {
-            this.operationState = OperationState.MessageCreated;
-
-            string responseLine = this.batchStream.ReadFirstNonEmptyLine();
-
-            int statusCode = this.ParseResponseLine(responseLine);
-
-            // Read all headers and create the response message
-            ODataBatchOperationHeaders headers = this.batchStream.ReadHeaders();
-
-            if (this.batchStream.ChangeSetBoundary != null)
-            {
-                if (this.allowLegacyContentIdBehaviour)
-                {
-                    // Add a potential Content-ID header to the URL resolver so that it will be available
-                    // to subsequent operations.
-                    string contentId;
-                    if (this.contentIdToAddOnNextRead == null && headers.TryGetValue(ODataConstants.ContentIdHeader, out contentId))
-                    {
-                        if (contentId != null && this.urlResolver.ContainsContentId(contentId))
-                        {
-                            throw new ODataException(Strings.ODataBatchReader_DuplicateContentIDsNotAllowed(contentId));
-                        }
-
-                        this.contentIdToAddOnNextRead = contentId;
-                    }
-                }
-            }
-
-            // In responses we don't need to use our batch URL resolver, since there are no cross referencing URLs
-            // so use the URL resolver from the batch message instead.
-            ODataBatchOperationResponseMessage responseMessage = ODataBatchOperationResponseMessage.CreateReadMessage(
-                this.batchStream,
-                statusCode,
-                headers,
-                this.contentIdToAddOnNextRead,
-                /*operationListener*/ this,
-                this.urlResolver.BatchMessageUrlResolver);
-
-            //// NOTE: Content-IDs for cross referencing are only supported in request messages; in responses
-            ////       we allow a Content-ID header but don't process it (i.e., don't add the content ID to the URL resolver).
-
-            return responseMessage;
+            ParseRequestLine(requestLine, /*httpMethodValidation*/ null, out httpMethod, out requestUri);
         }
 
         /// <summary>
         /// Parses the request line of a batch operation request.
         /// </summary>
         /// <param name="requestLine">The request line as a string.</param>
+        /// <param name="httpMethodValidation">The validation for the HTTP method in the request line.</param>
         /// <param name="httpMethod">The parsed HTTP method of the request.</param>
         /// <param name="requestUri">The parsed <see cref="Uri"/> of the request.</param>
-        private void ParseRequestLine(string requestLine, out string httpMethod, out Uri requestUri)
+        internal void ParseRequestLine(string requestLine, Action<string> httpMethodValidation, out string httpMethod, out Uri requestUri)

```

Comment:
>httpMethodValidation [](start = 74, length = 20)

If this is only called by ODataMultipartMixedBatchReader, and you move it there, can you inline this logic rather than pass it in as an action?